chatbot-core/rag/retriever/retriever_bm25.py

```diff
@@ -0,0 +1,70 @@
+"""
+Query interface for retrieving the most relevant embedded text chunks using a Sparse Retriever.
+"""
+
+from rag.retriever.retriever_utils import load_vector_index
+from rag.embedding.bm25_indexer import indexer
+from api.config.loader import CONFIG
+
+def perform_keyword_search(query, logger, source_name, top_k=5):
+    """
+    Retrieve the top-k most relevant chunks for a given natural language query
+    performing a keyword based search.
+
+    Args:
+        query (str): The input query string.
+        logger (logging.Logger): Logger for warnings and file-level updates.
+        source_name (str): The source name that we want to consider.
+        top_k (int): Number of top results to retrieve. Defaults to 5.
+
+    Returns:
+        tuple[list[dict], list[float]]: Retrieved data and similarity scores.
+    """
+    if not query.strip():
+        logger.warning("Empty query received.")
+        return [], []
+    index = indexer.get(source_name)
+    _, metadata = load_vector_index(logger, source_name)
+
+    if not index or not metadata:
+        return [], []
+
+    data, scores = search_bm25_index(query, index, metadata, logger, top_k)
+
+    filtered = [(d, s) for d, s in zip(data, scores)
+                if s >= CONFIG["retrieval"]["keyword_threshold"]]
```

Comment:
This line tightly couples the function to the global CONFIG object. For better testability and reusability, pass this value as an argument